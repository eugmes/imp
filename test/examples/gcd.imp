-- FIXME is_even is not allowed here
function iseven (a : integer) return boolean is
begin
  return (a / 2) * 2 == a;
end iseven;

function gcd (a, b : integer) return integer is
  var d : integer;
begin
  d = 1;

  while iseven (a) and iseven(b) loop
    a = a / 2;
    b = b / 2;
    d = d * 2;
  end loop;

  while a # b loop
    if iseven (a) then
      a = a / 2;
    elsif iseven (b) then
      b = b / 2;
    elsif a > b then
      a = (a - b) / 2;
    else
      b = (b - a) / 2;
    end if;
  end loop;

  return a * d;
end gcd;

procedure main () is
   var a, b, c : integer;
begin
   output ("Firsr number: "); input (a);
   output ("Second number: "); input (b);
   c = gcd (a, b);
   output (c); newline;
end main;
