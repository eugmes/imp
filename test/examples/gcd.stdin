4200
3528
