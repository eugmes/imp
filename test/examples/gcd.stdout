Firsr number: Second number: 168
